pkg/kubelet/dockershim/docker_sandbox.go

```diff
@@ -19,33 +19,243 @@ package dockershim
+		IP = r.NetworkSettings.IPAddress
+		// Fall back to IPv6 address if no IPv4 address is present
+		if IP == "" {
+			IP = r.NetworkSettings.GlobalIPv6Address
+		}
+	}
+	network := &runtimeApi.PodSandboxNetworkStatus{Ip: &IP}
+	netNS := getNetworkNamespace(r)
+
+	return &runtimeApi.PodSandboxStatus{
+		Id:        &r.ID,
+		Name:      &r.Name,
+		State:     &state,
+		CreatedAt: &ct,
+		// TODO: We write annotations as labels on the docker containers. All
+		// these annotations will be read back as labels. Need to fix this.
+		Labels:  r.Config.Labels,
+		Network: network,
+		Linux:   &runtimeApi.LinuxPodSandboxStatus{Namespaces: &runtimeApi.Namespace{Network: &netNS}},
+	}, nil
 }
 
 // ListPodSandbox returns a list of Sandbox.
 func (ds *dockerService) ListPodSandbox(filter *runtimeApi.PodSandboxFilter) ([]*runtimeApi.PodSandbox, error) {
-	return nil, fmt.Errorf("Not implemented")
+	// By default, list all containers whether they are running or not.
+	opts := dockertypes.ContainerListOptions{All: true}
+	filterOutReadySandboxes := false
+
+	opts.Filter = dockerfilters.NewArgs()
+	if filter != nil {
+		if filter.Name != nil {
+			opts.Filter.Add("name", filter.GetName())
+		}
+		if filter.Id != nil {
+			opts.Filter.Add("id", filter.GetId())
+		}
+		if filter.State != nil {
+			if filter.GetState() == runtimeApi.PodSandBoxState_READY {
+				// Only list running containers.
+				opts.All = false
+			} else {
+				// runtimeApi.PodSandBoxState_NOTREADY can mean the
+				// container is in any of the non-running state (e.g., created,
+				// exited). We can't tell docker to filter out running
+				// containers directly, so we'll need to filter them out
+				// ourselves after getting the results.
+				filterOutReadySandboxes = true
+			}
+		}
+
+		if filter.LabelSelector != nil {
+			for k, v := range filter.LabelSelector {
+				opts.Filter.Add(k, v)
+			}
+		}
+		// Filter out sandbox containers.
+		opts.Filter.Add(containerTypeLabelKey, containerTypeLabelSandbox)
+	}
+	containers, err := ds.client.ListContainers(opts)
+	if err != nil {
+		return err, nil
+	}
+
+	// Convert docker containers to runtime api sandboxes.
+	result := []*runtimeApi.PodSandbox{}
+	for _, c := range containers {
+		s := toRuntimeAPISandbox(&c)
+		if filterOutReadySandboxes && s.GetState() == runtimeApi.PodSandBoxState_READY {
+			continue
+		}
+		result = append(result, s)
+	}
+	return result, nil
+}
+
+func makeSandboxDockerConfig(c *runtimeApi.PodSandboxConfig, image string) *dockertypes.ContainerCreateConfig {
+	// Merge annotations and labels because docker supports only labels.
+	labels := makeLabels(c.GetLabels(), c.GetAnnotations())
+	// Apply a label to distinguish sandboxes from regular containers.
+	labels[containerTypeLabelKey] = containerTypeLabelSandbox
+
+	hc := &dockercontainer.HostConfig{}
+	createConfig := &dockertypes.ContainerCreateConfig{
+		Name: c.GetName(),
+		Config: &dockercontainer.Config{
+			Hostname: c.GetHostname(),
```

Comment:
Same question again, should we handle the host network logic at higher level or here.

Currently, we only set hostname, dns, dnssearch when the pod is not using host network.
We can handle this in either higher level or at specific runtime level.
* If we handle it at higher level, what should be passed here, empty value?
* If we handle it at this level, what should be passed here, pod hostname?